iamauthz/middleware.go

```diff
@@ -0,0 +1,73 @@
+package iamauthz
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Authorize the current request.
+// WithAuthorization must have been called on the context for the call to be effective.
+func Authorize(ctx context.Context) {
+	if value, ok := ctx.Value(contextKey{}).(*contextValue); ok {
+		value.mu.Lock()
+		value.authorized = true
+		value.mu.Unlock()
+	}
+}
+
+// RequireUnaryAuthorization is a grpc.UnaryServerInterceptor that requires authorization
+// to be performed on all incoming requests.
+func RequireUnaryAuthorization(
+	ctx context.Context,
+	req interface{},
+	_ *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	ctx = WithAuthorization(ctx)
+	resp, err := handler(ctx, req)
+	if code := status.Code(err); code == codes.Unauthenticated || code == codes.PermissionDenied {
+		return nil, err
+	}
+	value := ctx.Value(contextKey{}).(*contextValue)
+	value.mu.Lock()
+	authorized := value.authorized
+	value.mu.Unlock()
+	if !authorized {
+		return nil, status.Error(codes.Internal, "server did not perform authorization")
+	}
+	return resp, err
```

Comment:
Not sure how we should do with swallowing of errors (depends on where we put logging I suppose).
If my handler returns `Internal` early and is unable to perform authorization, the original error would be lost. 
Maybe a better example is if the authorization logic fails with `DeadlineExceeded` (reading from spanner), for which I guess we would return `Internal`.

Reply:
I think it should be OK to swallow errors here - if this middleware gets triggered, it should trigger an alert saying that some endpoint that should have authorization does not have it, and it should be a prioritized fix?

Reply:
For the second issue (authorization logic times out) the issue is not that authorization is not implemented, but rather network issues. This middleware would swallow the `DeadlineExceeded` issue and only report that authorization is not implemented.

Reply:
I see - we would need to call `iamauthz.Authorize(ctx)` at the top of the authorization, or not swallow errors. Do you think we should drop this PR and do something else?

Reply:
In that case I think call `Authorize` at the start is better, because the error messages *could*  include information that you need to be authorized to see.

Reply:
Good point! I updated the example implementation, and the documentation for the iamauthz package - what do you think?